### PR TITLE
feat: add invoke action

### LINKED ISSUE #96: Allow invoke feature for functions
Run `kn func invoke`expanding it to trigger both HTTP and CloudEvent Functions with detection of local vs remote targets.

Refer to this upstream [PR](https://github.com/knative-sandbox/kn-plugin-func/pull/705) for details. 

Previous command was `kn func emit` with details here https://github.com/knative-sandbox/kn-plugin-func/blob/main/docs/guides/commands.md#emit

### PR DESCRIPTION
**Pull request by lstocchi**
it resolves #96 

An example below. Few things to be mentioned.
1) If you open an only local/only remote function you won't see the instances options at the top.
2) The content-type textboxes is auto-filled based on the file picked by the user (e.g a file .jpg will have the image/jpeg mime type automatically selected)
3) The func cli claims to get values from env variables for their args. I only check those related to source/format/type/content-type/id. I skipped the others as i don't think it makes sense from an IDE perspective :thinking: 

invoke_func

invoke_autocomplete

cc @sudhirverma @mohitsuman

### CODE REVIEW

**Comment by mohitsuman** [comment]
@lstocchi This looks good. 👍

File: src/main/java/com/redhat/devtools/intellij/knative/kn/KnCli.java [hunk 1]
```diff
@@ -308,6 +309,48 @@ public void deployFunc(String namespace, String path, String registry, String im
         return args.toArray(new String[0]);
     }
 
+    @Override
+    public void invokeFunc(InvokeModel model) throws IOException {
+        ExecHelper.executeWithTerminal(project, KNATIVE_TOOL_WINDOW_ID, getInvokeArgs("invoke", model));
```
Comment (jeffmaury) [inline src/main/java/com/redhat/devtools/intellij/knative/kn/KnCli.java:314]: The `envVars` is missing so this is wrong on proxy enabled sites. I realized the whole KnCli is wrong when calling `executeWithTerminal` so maybe we should open a separate issue ?

File: src/main/java/com/redhat/devtools/intellij/knative/utils/MimeTypes.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
```
Comment (jeffmaury) [inline src/main/java/com/redhat/devtools/intellij/knative/utils/MimeTypes.java:1]: Should be in IJ common or use an API from the IJ SDK (FileType ?)

**Comment by jeffmaury** [comment]
Just tried on my local function but got no UI and go the following exception in the log:

```
java.lang.NullPointerException
	at com.redhat.devtools.intellij.knative.ui.invokeFunc.InvokeDialog.addInstanceButtonListener(InvokeDialog.java:195)
	at com.redhat.devtools.intellij.knative.ui.invokeFunc.InvokeDialog.fillContentPanel(InvokeDialog.java:144)
	at com.redhat.devtools.intellij.knative.ui.invokeFunc.InvokeDialog.buildStructure(InvokeDialog.java:116)
	at com.redhat.devtools.intellij.knative.ui.invokeFunc.InvokeDialog.<init>(InvokeDialog.java:90)
	at com.redhat.devtools.intellij.knative.actions.func.InvokeAction.actionPerformed(InvokeAction.java:54)
	at com.redhat.devtools.intellij.knative.actions.KnAction.actionPerformed(KnAction.java:63)
	at com.redhat.devtools.intellij.knative.actions.KnAction.actionPerformed(KnAction.java:50)
	at com.redhat.devtools.intellij.common.actions.TreeAction.actionPerformed(TreeAction.java:127)
	at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:266)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:283)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.lambda$actionPerformed$0(ActionMenuItem.java:294)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.runOnOwnContext(FocusManagerImpl.java:263)
	at com.intellij.openapi.wm.impl.IdeFocusManagerImpl.runOnOwnContext(IdeFocusManagerImpl.java:77)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.actionPerformed(ActionMenuItem.java:284)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.lambda$fireActionPerformed$0(ActionMenuItem.java:112)
	at com.intellij.openapi.application.TransactionGuardImpl.runSyncTransaction(TransactionGuardImpl.java:83)
	at com.intellij.openapi.application.TransactionGuardImpl.lambda$submitTransaction$1(TransactionGuardImpl.java:107)
	at com.intellij.openapi.application.TransactionGuardImpl.submitTransaction(TransactionGuardImpl.java:116)
	at com.intellij.openapi.application.TransactionGuard.submitTransaction(TransactionGuard.java:121)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.fireActionPerformed(ActionMenuItem.java:112)
	at com.intellij.ui.plaf.beg.BegMenuItemUI.doClick(BegMenuItemUI.java:517)
	at com.intellij.ui.plaf.beg.BegMenuItemUI.access$300(BegMenuItemUI.java:36)
	at com.intellij.ui.plaf.beg.BegMenuItemUI$MyMouseInputHandler.mouseReleased(BegMenuItemUI.java:539)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6651)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6416)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5026)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4858)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4858)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:906)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:844)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:776)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:422)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:698)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:421)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by lstocchi** [comment]
> Just tried on my local function but got no UI and go the following exception in the log:
> 

Sorry, when cleaning the code i forgot to add a check for NPE and i was testing with a function which was both local/remote  so i never faced it. Now it should be fixed.

**Comment by lstocchi** [comment]
I just noticed that I missed to add a field for url when targeting a remote instance

File: src/main/java/com/redhat/devtools/intellij/knative/actions/func/InvokeAction.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        if (dialog.isOK()) {
+            ExecHelper.submit(() -> {
+                try {
+                    knCli.invokeFunc(model);
+                    telemetry
+                            .result(anonymizeResource(name, namespace, "Invoked function " + name))
+                            .send();
+                } catch (IOException e) {
+                    logger.warn(e.getLocalizedMessage());
```
Comment (jeffmaury) [inline src/main/java/com/redhat/devtools/intellij/knative/actions/func/InvokeAction.java:72]: The exception is object is missing from the method parameters, it is useful as the whole stacktrace will be added to the log if you pass it.

**Review by jeffmaury — CHANGES_REQUESTED** [review]

**Pull request merged**